examples/graph/gnn_citations.py

```diff
@@ -0,0 +1,728 @@
+"""
+Title: Node Classification with Graph Neural Networks
+Author: [Khalid Salama](https://www.linkedin.com/in/khalid-salama-24403144/)
+Date created: 2021/05/30
+Last modified: 2021/05/30
+Description: Implementing a graph neural network model for predicting the topic of a paper given its citations.
+"""
+
+"""
+## Introduction
+
+Many datasets in various machine learning (ML) applications have structural relationships
+between their entities, which can be represented as graphs. Such application includes
+social and communication networks analysis, traffic prediction, and fraud detection.
+[Graph representation Learning](https://www.cs.mcgill.ca/~wlh/grl_book/)
+aims to build and train models for graph datasets to be used for a variety of ML tasks.
+
+This example demonstrate a simple implementation of a [Graph Neural Network](https://arxiv.org/pdf/1901.00596.pdf)
+(GNN) model. The model is used for a node prediction task on the [Cora dataset](https://relational.fit.cvut.cz/dataset/CORA)
+to predict the subject of a paper given its words and citations network.
+
+Note that, **we implement a Graph Convolution Layer from scratch** to provide better
+understanding of how they work. However, there is a number of specialized TensorFlow-based
+libraries that provide rich GNN APIs, such as [Spectral](https://graphneural.network/),
+[StellarGraph](https://stellargraph.readthedocs.io/en/stable/README.html), and
+[GraphNets](https://github.com/deepmind/graph_nets).
+"""
+
+"""
+## Setup
+"""
+
+import os
+import tarfile
+from urllib.request import urlretrieve
+import pandas as pd
+import numpy as np
+import networkx as nx
+import matplotlib.pyplot as plt
+import tensorflow as tf
+from tensorflow import keras
+from tensorflow.keras import layers
+
+"""
+## Prepare the Dataset
+
+The Cora dataset consists of 2,708 scientific papers classified into one of seven classes.
+The citation network consists of 5,429 links. Each paper has a binary word vector of size
+1,433, indicating the presence of a corresponding word.
+
+### Download the dataset
+
+The dataset has two tap-separated files: `cora.cites` and `cora.content`.
+
+1. The `cora.cites` includes the citation records with two columns:
+`cited_paper_id` (target) and `citing_paper_id` (source).
+2. The `cora.content` includes the paper content records with 1,435 columns:
+`paper_id`, `subject`, and 1,433 binary features.
+
+Let's download the dataset.
+"""
+
+urlretrieve("https://linqs-data.soe.ucsc.edu/public/lbc/cora.tgz", "cora.tgz")
```

Comment:
Have you tried using `keras.utils.get_file(url, extract=True)`?

Reply:
Done